lmdeploy/pytorch/kernels/cuda/pagedattention.py

```diff
@@ -243,6 +245,310 @@ def _fwd_grouped_split_kernel(
     tl.store(Acc_out + off_meta + 1, l_i, mask=mask_h)
 
 
+@triton.jit
+def _unpack_kv_int4(k):
```

Comment:
Have you tried loading k with `offs=[0, 0, 1, 1, 2, 2, ...]` ? unpacking would be easy with shift and mask.

Reply:
Do you mean loading k with `offs=[0,1,...,len, 0, 1,...,len]`? But we still have to split the tensor. Or we can load k twice.